test/EndToEndTests/Tests/Client/Microsoft.OData.Client.E2E.Tests/PayloadValueConverterTests/PayloadValueConverterTests.cs

```diff
@@ -0,0 +1,94 @@
+
+using Microsoft.AspNetCore.OData;
+using Microsoft.AspNetCore.OData.Routing.Controllers;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.OData.Client.E2E.Tests.PayloadValueConverterTests.Client.Default;
+using Microsoft.OData.Client.E2E.Tests.PayloadValueConverterTests.Server;
+using Microsoft.OData.E2E.TestCommon;
+using Microsoft.OData.E2E.TestCommon.Common;
+using Microsoft.OData.E2E.TestCommon.Common.Server.Default;
+using Microsoft.OData.E2E.TestCommon.Converters;
+using Microsoft.OData.Edm;
+using Xunit;
+using ClientPayloadConverterModel = Microsoft.OData.Client.E2E.Tests.PayloadValueConverterTests.Client;
+
+namespace Microsoft.OData.Client.E2E.Tests.PayloadValueConverterTests
+{
+    public class PayloadValueConverterTests : EndToEndTestBase<PayloadValueConverterTests.TestsStartup>
+    {
+        private readonly Uri _baseUri;
+        private readonly Container _context;
+        private readonly IEdmModel _model;
+
+        public class TestsStartup : TestStartupBase
+        {
+            public override void ConfigureServices(IServiceCollection services)
+            {
+                services.ConfigureControllers(typeof(PayloadValueConverterController), typeof(MetadataController));
+
+                services.AddControllers().AddOData(opt => opt.Count().Filter().Expand().Select().OrderBy().SetMaxTop(null)
+                    .AddRouteComponents(
+                    "odata",
+                    PayloadValueConverterEdmModel.GetEdmModel(),
+                    a => a.AddSingleton<ODataPayloadValueConverter, BinaryPayloadConverter>()));
+            }
+        }
+
+        public PayloadValueConverterTests(TestWebApplicationFactory<TestsStartup> fixture)
+            : base(fixture)
+        {
+            _baseUri = new Uri(Client.BaseAddress, "odata/");
+
+            _context = new Container(_baseUri)
+            {
+                HttpClientFactory = HttpClientFactory
+            };
+
+            _model = DefaultEdmModel.GetEdmModel();
+            ResetDataSource();
+        }
+
+        [Fact]
+        public async Task PostingAndQueryingABinaryProperty_UsingCustomBinaryToHyphenStringConverter_ShouldReturnHyphenSeparatedString()
+        {
+            var person = new ClientPayloadConverterModel.Person
+            {
+                Picture = new byte[] { 3, 1, 4 },
+                BusinessCard = new ClientPayloadConverterModel.ContactInfo()
+                {
+                    N = "T"
+                }
+            };
+
+            _context.AddToPeople(person);
+            await _context.SaveChangesAsync();
+
+            var requestUrl = new Uri(_baseUri.AbsoluteUri + "People(" + person.Id + ")/Picture", UriKind.Absolute);
+
+            var requestMessage = new TestHttpClientRequestMessage(requestUrl, base.Client)
+            {
+                Method = "GET"
+            };
+
+            var responseMessage = await requestMessage.GetResponseAsync();

```

Comment:
It seems that we're not using the OData Client API to read this response, we're using `TestHttpClientRequestMessage`, which means we might as well have used the plain `HttpClient`, again this makes me wonder why this has to be an E2E test since it doesn't seem like any relevant OData client behaviour is being verified here. It seems like what we're really verifying is whether AspNetCoreOData supports a custom converter and uses it to process the result. Is that the case?

Is it that it would be difficult to test this in AspNetCoreOData and it's easier to test in an E2E test?

Reply:
The `ODataPayloadValueConverter` that is being overriden is a class in `Microsoft.OData.Core` but I get what you're saying.